pandas/core/dtypes/cast.py

```diff
@@ -264,74 +278,390 @@ def changeit():
-            dtype = np.float64
-    elif is_bool(fill_value):
-        if not issubclass(dtype.type, np.bool_):
-            dtype = np.object_
-    elif is_integer(fill_value):
-        if issubclass(dtype.type, np.bool_):
-            dtype = np.object_
-        elif issubclass(dtype.type, np.integer):
-            # upcast to prevent overflow
-            arr = np.asarray(fill_value)
-            if arr != arr.astype(dtype):
-                dtype = arr.dtype
-    elif is_complex(fill_value):
-        if issubclass(dtype.type, np.bool_):
-            dtype = np.object_
-        elif issubclass(dtype.type, (np.integer, np.floating)):
-            dtype = np.complex128
-    elif fill_value is None:
-        if is_float_dtype(dtype) or is_complex_dtype(dtype):
-            fill_value = np.nan
-        elif is_integer_dtype(dtype):
-            dtype = np.float64
-            fill_value = np.nan
-        elif is_datetime_or_timedelta_dtype(dtype):
-            fill_value = iNaT
-        else:
-            dtype = np.object_
-            fill_value = np.nan
+    Parameters
+    ----------
+    dtype : DType
+        The dtype to start from.
+    fill_value : scalar or np.ndarray / Series / Index
+        The value that the output dtype needs to be able to hold.
+
+    Returns
+    -------
+    dtype : DType
+        The updated dtype.
+    fill_value : scalar
+        The type of this value depends on the type of the passed fill_value
+
+        * If fill_value is a scalar, the method returns that scalar, but
+          modified to fit the updated dtype. For example, a datetime fill_value
+          will be returned as an integer (representing ns) for M8[ns], and
+          values considered missing (see pd.isna) will be returned as the
+          corresponding missing value marker for the updated dtype.
+        * If fill_value is an ndarray/Series/Index, this method will always
+          return the missing value marker for the updated dtype. This value
+          will be None for dtypes that cannot hold missing values (integers,
+          booleans, bytes).
+
+    See Also
+    --------
+    maybe_promote_with_scalar : underlying method for scalar case
+    maybe_promote_with_array : underlying method for array case
+    """
+    if is_scalar(fill_value) or isinstance(fill_value, tuple):
+        return maybe_promote_with_scalar(dtype, fill_value)
+    elif isinstance(fill_value, (np.ndarray, ABCSeries, ABCIndexClass)):
+        return maybe_promote_with_array(dtype, fill_value)
     else:
-        dtype = np.object_
+        fill_type = type(fill_value).__name__
+        raise ValueError('fill_value must either be scalar, or a Series / '
+                         'Index / np.ndarray; received {}'.format(fill_type))
 
-    # in case we have a string that looked like a number
-    if is_extension_array_dtype(dtype):
-        pass
-    elif is_datetime64tz_dtype(dtype):
+
+def maybe_promote_with_scalar(dtype, fill_value=np.nan):
```

Comment:
this and the array variant are supposed to be called only from maybe_promote right?  If so, privatize.

Reply:
I can privatize them now, but the idea is actually to (later) replace the call-sites of `maybe_promote` with the correspondingly appropriate call; since the behaviour between `_with_scalar` and `_with_array` is quite different.